### ISSUE TITLE
"mounting" subrouters

### ISSUE THREAD

**Issue by timoxley** [yoshuawuyts/wayfarer#12]
I'd initially expected subrouters to automatically nest inside their parent routes like so:

```js
const parent = wayfarer()
const child = wayfarer()

parent.on('/parent', child)
child.on('/child', () => {
  console.log('hello')
})

parent('/parent/child') // => 'hello'
```

Perhaps could add a `mount` method that's similar to the `on` method, except that it does some additional work to enable the above? Or is this adding too much magic?

**Comment by yoshuawuyts** [comment]
Ah yes, this also bugged me. I initially wanted to let `.on()` do that by default, but figured it might yield unexpected results. Glad you opened up an issue about it.

### issues
There's actually two issues you're outlining in your example:
- subrouters need to be aware of the parents' prefix
- parent routers don't match arbitrary child routes (e.g. `/parent` doesn't match `/parent/child`)

### constraints
Any possible solutions will have the following constraints:
- earlier matched params must be preserved
- routington [doesn't allow glob matching](https://github.com/pillarjs/routington/issues/24)
- preferably we don't want to expand the api surface

### solution?
I think `.on()` can actually be extended to automatically detect subrouters and act as expected. My greatest concern is if there's a legitimate use case where paths shouldn't be trimmed for subrouters, but I can't really think of one (as long as matched params are passed through).

- extend `.on(path)` to `.on(path, params)`
- whenever `.on()` detects an instance of wayfarer (using [symbols](https://gist.github.com/yoshuawuyts/2bf8d5394e6f995791a0#file-symbol-function-id-js)), delegate to an internal `mount` fn 
- internal `mount` matches both `/parent` and `/parent/child` (trim passed in string to match)
- internal `mount` removes the leading path when passed into the subrouter (`/parent/child` becomes `/child`)

```js
const parent = wayfarer()
const child = wayfarer()

parent.on('/:parent', child)
child.on('/child', (uri, param) => {
  console.log('uri: ', uri)
  console.log('param: ', param)
})

parent('/dada/child') 
// => 'uri: /child'
// => 'param: {parent: dada}'
```

Guess I'm going a different way than what you proposed. Not sure if we cover all edge cases with this.  Let me know what you think!

**Comment by timoxley** [comment]
that all sounds good.

Note I believe you can use the new "Symbol.species" key in a similar manner to how you're using symbols there. http://www.2ality.com/2015/02/es6-classes-final.html#the_species_pattern

**Comment by timoxley** [comment]
Also just thinking… what if you pass a router from one instance of `wayfarer` to another… e.g. from a different `node_modules/wayfarer` instance… it's possible they're even the same version, yet they won't detect compatibility because the Symbol will be unique to each `node_modules/wayfarer` instance.

**Comment by timoxley** [comment]
Perhaps restricting to only mounting instances from the same wayfarer is good enough. What I'm describing is probably an edge case, but I'm interested in how to solve this problem in general.

**Comment by timoxley** [comment]
perhaps could use the global symbol registry… drop a global symbol in for each major version of a package.

**Comment by yoshuawuyts** [comment]
Wasn't aware of [`Symbol.for()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol/for), it indeed looks like a good solution for this problem. Don't think depending on a global scope is harmful here, especially since it's constraining itself to its package name.

`Symbol.species()` looks like a more generic solution though; I kind of dig it (since it doesn't rely on globals and is inspectable). If I understand correctly, the following should then be possible:

```js
const wayfarer = require('wayfarer')

const r1 = wayfarer()
console.log(r1 instanceof wayfarer)
// => true
```

though I'm not sure if this works in an ES5 environment, [`es6-symbol`](https://github.com/medikoo/es6-symbol) probably needs testing.

**Comment by timoxley** [comment]
:+1:

**Comment by yoshuawuyts** [comment]
Neither `Symbol.species` or `Symbol.for` worked well with `es6-symbol`, I don't think the `instanceof` call quite knows how to handle them. Created a wip branch; think I'm close to hitting the mark, but super sleepy so I'll finish it another time.

**Comment by timoxley** [comment]
I'd suggest getting some babel going, or perhaps just core-js?

**Comment by yoshuawuyts** [comment]
I'd rather not pull in babel unless absolutely forced to do so; marking functions with `es6-symbol` seems to work reasonably well so I think it's good to stick with it for backwards compat.

**Comment by timoxley** [comment]
aw, why?

**Comment by yoshuawuyts** [comment]
I haven't found a good way to pull in babel for modules; if you have any I'd be glad to hear. I'd rather not require on polyfills for modules to work / have an extra file that looks like:
```js
// enable babel in subsequent require calls
require('babel')

// load main code
require('./index')
```

**Comment by yoshuawuyts** [comment]
PR is ready! :tada: - https://github.com/yoshuawuyts/wayfarer/pull/14

**Comment by yoshuawuyts** [comment]
published as `wayfarer@4.0.0`
